Add Quick-start integration testing guide (#26)
ansible/community-docs: integration_tests_quick_start_guide.rst

.. code:: yaml

  - name: Create PostgreSQL user and store module's output to the result variable
    postgresql_user:
      name: test_user
    register: result

  - name: Check the module returns what we expect
    assert:
      that:
        - result is changed

  - name: Check actual system state with another module, in other words, that the user exists
    postgresql_query:
      query: SELECT * FROM pg_authid WHERE rolename = 'test_user'
    register: query_result

  - name: We expect it returns one row, check it
    assert:
      that:
        - query_result.rowcount == 1

You will learn details in the following sections.

The basic entity of Ansible integration tests is a ``target``.

The target is an `Ansible role <https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html>`_ stored in the ``tests/integration/targets`` directory of a collection repository file tree.

The target role contains all needed to test a module.

Names of targets contain a module name they test.

> Target names that start with ``setup_`` are executed as dependencies before module and plugin targets start execution. We will describe this kind of targets later in detail in the :ref:`Writing tests from scratch<Writing-tests-from-scratch>` section.

Reviewer: This is a convention, not a hard requirement, so let's better write:
```suggestion
Target names that start with ``setup_`` are usually executed as dependencies before module and plugin targets start execution. We will describe this kind of targets later in detail in the :ref:`Writing tests from scratch<Writing-tests-from-scratch>` section.
```

To run integration tests, we use the ``ansible-test`` utility shipped with the ``ansible-core`` and ``ansible`` packages. This is described in the :ref:`Run integration tests<Run-integration-tests>` section.

After you finish your integration tests, refert to the `Create a PR quick-start guide <https://github.com/ansible/community-docs/blob/main/create_pr_quick_start_guide.rst>`_ to learn how to submit a pull request.

.. _Prepare-local-environment:

Prepare local environment
=========================

Before starting working on integration tests, to be able to run them locally, you need to prepare your environment.

To learn how to do it quickly, refer to the `Quick-start development guide <https://github.com/ansible/community-docs/blob/main/create_pr_quick_start_guide.rst#prepare-your-environment>`_.

> .. _Determine-if-integration-tests-exists

Reviewer: ```suggestion
.. _Determine-if-integration-tests-exists:
```

If you already have your local environment :ref:`prepared<Prepare-local-environment>`, you can run the following command being in the collection's root directory to list all the available targets:

.. code:: bash

  ansible-test integration --list-targets

If you use ``bash`` and the ``argcomplete`` package is installed via ``pip`` on your system, you can also get a full target list by doing: ``ansible-test integration <tab><tab>``.
Alternatively, you can check if the ``tests/integration/targets`` contains a corresponding directory named as the module.

For example, the tests for the ``postgresql_user`` module of the ``community.postgresql`` collection are stored in the ``tests/integration/targets/postgresql_user`` directory of the collection's source tree.

If there is no corresponding target there, it means that the module does not have integration tests. In this case, think of adding integration tests for the module. Refer to the :ref:`Writing tests from scratch<Writing-tests-from-scratch>` section for details.

.. _Adding-tests-to-existing-ones:

Adding your tests to existing ones
==================================

> The test tasks are stored in the ``tests/integration/targets/<target_name>/tasks`` directory.

Reviewer: Do you also want to mention runme.sh based tests? I think it's worth at least mentioning them, so readers of this document aren't completely surprised when they see them.

Author: I've no idea what is this:))) I can see this mentioned briefly only once in `ansible/docs/docsite/rst/dev_guide/developing_collections_testing.rst`.
Is it an alternative to `setup_*` targets?
If you know how it works, could you please suggest good explanation for that?:) Would be really interesting to learn the  alternative concept.

Author: added, thanks:)

    when: postgres_version_resp.stdout is version('9.4', '>=')

We will add the following code to the file.

.. code:: yaml

  # https://github.com/ansible-collections/community.postgresql/issues/NUM
  - name: Test user name containing underscore
    postgresql_user:
      name: underscored_user
    register: result

  - name: Check the module returns what we expect
    assert:
      that:
        - result is changed

  - name: Query the database if the user exists
    postgresql_query:
      query: SELECT * FROM pg_authid WHERE rolename = 'underscored_user'
    register: result

  - name: Check the database returns one row
    assert:
      that:
        - query_result.rowcount == 1

When we :ref:`run the tests<Run-integration-tests>` passing ``postgresql_user`` as a test target, this task must fail.

Then we will fix the bug and run the same tests again. If they pass, we will consider the bug fixed and will submit a pull request.

When adding a new feature
-------------------------

.. note::

  The process described in this section is also applicable when a feature already exists but does not have integration tests and you want to cover it.

.. note::

>   If you don not implement the feature you want yet, you can start with writing integration tests for it. Of course they will not work as the code does not exist at the moment but it can help you design better implementation before writing the code.

Reviewer: ```suggestion
  If you do not implement the feature you want yet, you can start with writing integration tests for it. Of course they will not work as the code does not exist at the moment but it can help you design better implementation before writing the code.
```

  - name: Test for new_option, create new user WITHOUT the attribute
    postgresql_user:
      name: test_user
      add_attribute: no
    register: result

  - name: Check the module returns what we expect
    assert:
      that:
        - result is changed

  - name: Query the database if the user exists but does not have the attribute (it is NULL)
    postgresql_query:
      query: SELECT * FROM pg_authid WHERE rolename = 'test_user' AND attribute = NULL
    register: result

  - name: Check the database returns one row
    assert:
      that:
        - query_result.rowcount == 1

  - name: Test for new_option, create new user WITH the attribute
    postgresql_user:
      name: test_user
      add_attribute: yes
    register: result

  - name: Check the module returns what we expect
    assert:
      that:
        - result is changed

  - name: Query the database if the user has the attribute (it is TRUE)
    postgresql_query:
      query: SELECT * FROM pg_authid WHERE rolename = 'test_user' AND attribute = 't'
    register: result

  - name: Check the database returns one row
    assert:
      that:
        - query_result.rowcount == 1

> When we :ref:`run the tests<Run-integration-tests>` with ``postgresql_user`` passed as a test target.

Author: I think `Then` was intended instead of `When`:) thanks for spotting!

> In real world, we would alternate the tasks above with the same tasks run with the ``check_mode: yes`` option to be sure our option works as expected in check-mode as well. Refer to the :ref:`Recommendations on coverage<Recommendations-on-coverage>` section for details.

Reviewer 2: ```suggestion
In reality, we would alternate the tasks above with the same tasks run with the ``check_mode: yes`` option to be sure our option works as expected in check-mode as well. Refer to the :ref:`Recommendations on coverage<Recommendations-on-coverage>` section for details.
```
